File: Medium/130. Surrounded Regions/solution.java
/**
 * TC: O(n * m)
 * SC: 
 */

class Solution {
    int[][] dirs = {{1, 0}, {0, 1}, {0, -1}, {-1, 0}};
    public void solve(char[][] board) {
        
        if (board == null || board.length == 0 || board[0].length == 0) {
            return;
        }
        int rows = board.length;
        int cols = board[0].length;
        for (int i = 0; i < rows; i++) {
            if ( board[i][0] == 'O') {
                dfs(board, i, 0);
            }
            if (board[i][cols - 1] == 'O') {
                dfs(board, i, cols - 1);
            }
        }
        for (int j = 0; j < cols; j++) {
            if (board[0][j] == 'O') {
                dfs(board, 0, j);
            }
            if (board[rows - 1][j] == 'O') {
                dfs(board, rows - 1, j);
            }
        }
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (board[i][j] == 'O') {
                    board[i][j] = 'X';
                }
                if (board[i][j] == '*') {
                    board[i][j] = 'O';
                }
            }
        }
        return;
        
    }
    private void dfs(char[][] board, int x, int y) {
        
        board[x][y] = '*';
        for (int[] dir: dirs) {
            int newx = x + dir[0];
            int newy = y + dir[1];
            if (inArea(board, newx, newy)) {
                if (board[newx][newy] == 'O') {
                    dfs(board, newx, newy);
                }
            }
        }
    }
    private boolean inArea(char[][] board, int x, int y) {
        
        int rows = board.length;
        int cols = board[0].length;
        return x >= 0 && x < rows && y >= 0 && y < cols;
    }
}

File: Medium/130. Surrounded Regions/solution.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        
        if not board or len(board) == 0 or len(board[0]) == 0:
            return
        
        rows = len(board)
        cols = len(board[0])
        for i in range(rows):
            if board[i][0] == "O":
                self.dfs(board, i, 0)
        for i in range(rows):
            if board[i][cols - 1] == 'O':
                self.dfs(board, i, cols - 1)
        for i in range(cols):
            if board[0][i] == 'O':
                self.dfs(board, 0, i)
        for i in range(cols):
            if board[rows - 1][i] == 'O':
                self.dfs(board, rows - 1, i)
        
        for i in range(rows):
            for j in range(cols):
                if board[i][j] == "*":
                    board[i][j] = "O"
                elif board[i][j] == "O":
                    board[i][j] = "X"
        return
    
    def dfs(self, board: List[List[str]], x: int, y: int) -> None:
        
        board[x][y] = "*"
        dirs = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        rows = len(board)
        cols = len(board[0])
        
        for dir in dirs:
            newx = x + dir[0]
            newy = y + dir[1]
            if newx >= 0 and newx < rows and newy >= 0 and newy < cols:
                if board[newx][newy] == 'O':
                    self.dfs(board, newx, newy)
                